drivers/adc/adc_gecko.c

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright (c) 2019 Interay Solutions B.V.
+ * Copyright (c) 2019 Oane Kingma
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+
+#include <adc.h>
+#include <device.h>
+#include <kernel.h>
+#include <init.h>
+#include <soc.h>
+#include <em_adc.h>
+#include <em_cmu.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_gecko);
+
+#define ADC_PREFIX cmuClock_ADC
+#define CLOCK_ADC(id) _CONCAT(ADC_PREFIX, id)
+
+/* Acquisition times */
+static const u16_t acq_time_single_tbl[10] = {
+1, 2, 3, 4, 8, 16, 32, 64, 128, 256 };
+
+/* 16 external channels. */
+#define GECKO_CHANNEL_COUNT             32
+
+struct adc_gecko_ch_cfg {
+	ADC_Ref_TypeDef reference;
+	ADC_AcqTime_TypeDef acq_time;
+	ADC_PosSel_TypeDef input_positive;
+	bool initialized;
+};
+
+struct adc_gecko_data {
+	struct adc_context ctx;
+	struct device *dev;
+	u16_t *buffer;
+	u16_t *repeat_buffer;
+	u8_t channel_count;
+	struct adc_gecko_ch_cfg ch_cfg[GECKO_CHANNEL_COUNT];
+};
+
+struct adc_gecko_cfg {
+	ADC_TypeDef *base;
+	CMU_Clock_TypeDef clock;
+	u32_t prescaler;
+	void (*irq_cfg_func)(void);
+};
+
+#define DEV_DATA(dev) \
+	((struct adc_gecko_data *)(dev)->driver_data)
+#define DEV_CFG(dev) \
+	((struct adc_gecko_cfg *)(dev)->config->config_info)
+
+static int adc_gecko_check_buffer_size(const struct adc_sequence *sequence,
+				       u8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(u16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+			sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_gecko_check_resolution(const struct adc_sequence *sequence)
+{
+	int value = 0;
+
+	switch (sequence->resolution) {
+	case 6:
+		value = adcRes6Bit;
+		break;
+	case 8:
+		value = adcRes8Bit;
+		break;
+	case 12:
+		value = adcRes12Bit;
+		break;
+	default:
+		LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
+
+	return value;
+}
+
+static int start_read(struct device *dev, const struct adc_sequence *sequence)
+{
+	const struct adc_gecko_cfg *config = DEV_CFG(dev);
+	struct adc_gecko_data *data = DEV_DATA(dev);
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+	ADC_InitSingle_TypeDef sInit = ADC_INITSINGLE_DEFAULT;
+	struct adc_gecko_ch_cfg *config_ch = NULL;
+	u8_t index;
+	int res;
+
+	/* Only 1 channel supported, verify at least and only 1 is enabled */
+	if ((sequence->channels & (sequence->channels - 1)) != 0) {
+		LOG_ERR("Invalid selection of channels");
+		return -EINVAL;
+	}
+
+	/* Find designated channel */
+	index = find_lsb_set(sequence->channels) - 1;
+	config_ch = (struct adc_gecko_ch_cfg *)&data->ch_cfg[index];
```

Comment:
No need for this casting (here and in `adc_gecko_channel_setup` and `adc_gecko_init`).